File: app/utils/password_strength.py
"""
密码强度验证工具
"""
import re

def check_password_strength(password):
    """
    检查密码强度
    
    Args:
        password: 密码字符串
        
    Returns:
        tuple: (strength_level, score, message)
            strength_level: 'low', 'middle', 'high'
            score: 0-100 的分数
            message: 强度描述
    """
    if not password:
        return 'low', 0, '密碼不能為空'
    
    length = len(password)
    score = 0
    
    # 长度评分 (0-40分)
    if length >= 12:
        score += 40
    elif length >= 10:
        score += 30
    elif length >= 8:
        score += 20
    elif length >= 6:
        score += 10
    else:
        score += 0
    
    # 包含小写字母 (0-15分)
    if re.search(r'[a-z]', password):
        score += 15
    
    # 包含大写字母 (0-15分)
    if re.search(r'[A-Z]', password):
        score += 15
    
    # 包含数字 (0-15分)
    if re.search(r'\d', password):
        score += 15
    
    # 包含特殊字符 (0-15分)
    if re.search(r'[!@#$%^&*()_+\-=\[\]{};:\'",.<>?/\\|`~]', password):
        score += 15
    
    # 判断强度级别
    if score >= 70:
        return 'high', score, '強密碼'
    elif score >= 45:
        return 'middle', score, '中等密碼'
    else:
        return 'low', score, '弱密碼'


def validate_password_strength(password, min_strength='middle'):
    """
    验证密码是否满足最低强度要求
    
    Args:
        password: 密码字符串
        min_strength: 最低强度要求 ('low', 'middle', 'high')
        
    Returns:
        tuple: (is_valid, error_message)
    """
    strength_levels = {'low': 0, 'middle': 1, 'high': 2}
    
    strength, score, message = check_password_strength(password)
    
    if strength_levels.get(strength, 0) < strength_levels.get(min_strength, 1):
        requirements = []
        
        if min_strength == 'middle':
            requirements = [
                '密碼長度至少 8 個字符',
                '包含大小寫字母',
                '包含數字'
            ]
        elif min_strength == 'high':
            requirements = [
                '密碼長度至少 10 個字符',
                '包含大小寫字母',
                '包含數字',
                '包含特殊字符'
            ]
        
        return False, f'密碼強度不足（當前：{message}），需要滿足：' + '、'.join(requirements)
    
    return True, ''


def get_password_strength_details(password):
    """
    获取密码强度详细信息（供前端显示）
    
    Args:
        password: 密码字符串
        
    Returns:
        dict: 包含 strength, score, message, requirements 的字典
    """
    strength, score, message = check_password_strength(password)
    
    requirements = {
        'length': len(password) >= 8,
        'lowercase': bool(re.search(r'[a-z]', password)),
        'uppercase': bool(re.search(r'[A-Z]', password)),
        'number': bool(re.search(r'\d', password)),
        'special': bool(re.search(r'[!@#$%^&*()_+\-=\[\]{};:\'",.<>?/\\|`~]', password))
    }
    
    return {
        'strength': strength,
        'score': score,
        'message': message,
        'requirements': requirements
    }

File: static/js/password_strength.js
/**
 * 密碼強度檢測 JavaScript 版本
 */

function checkPasswordStrength(password) {
    if (!password) {
        return {
            valid: false,
            strength: 'low',
            score: 0,
            message: '請輸入密碼',
            checks: {}
        };
    }
    
    let score = 0;
    const checks = {
        length: false,
        uppercase: false,
        lowercase: false,
        digit: false,
        special: false
    };
    
    // 1. 長度檢查
    if (password.length >= 8) {
        checks.length = true;
        score += 20;
    }
    if (password.length >= 12) score += 10;
    if (password.length >= 16) score += 10;
    
    // 2. 大寫字母
    if (/[A-Z]/.test(password)) {
        checks.uppercase = true;
        score += 15;
        if ((password.match(/[A-Z]/g) || []).length >= 2) score += 5;
    }
    
    // 3. 小寫字母
    if (/[a-z]/.test(password)) {
        checks.lowercase = true;
        score += 15;
        if ((password.match(/[a-z]/g) || []).length >= 2) score += 5;
    }
    
    // 4. 數字
    if (/\d/.test(password)) {
        checks.digit = true;
        score += 15;
        if ((password.match(/\d/g) || []).length >= 2) score += 5;
    }
    
    // 5. 特殊符號
    if (/[!@#$%^&*(),.?":{}|<>_\-+=\[\]\\\/~`]/.test(password)) {
        checks.special = true;
        score += 15;
        if ((password.match(/[!@#$%^&*(),.?":{}|<>_\-+=\[\]\\\/~`]/g) || []).length >= 2) score += 10;
    }
    
    // 判斷強度等級
    let strength, valid, message, color;
    
    if (score >= 80) {
        strength = 'high';
        valid = true;
        message = '強度：高';
        color = 'success';
    } else if (score >= 60) {
        strength = 'middle';
        valid = true;
        message = '強度：中（可接受）';
        color = 'warning';
    } else {
        strength = 'low';
        valid = false;
        message = '強度：低（不符合要求）';
        color = 'danger';
    }
    
    return {
        valid: valid,
        strength: strength,
        score: score,
        message: message,
        color: color,
        checks: checks
    };
}

function renderPasswordStrengthMeter(inputId, meterId) {
    const input = document.getElementById(inputId);
    const meter = document.getElementById(meterId);
    
    if (!input || !meter) return;
    
    input.addEventListener('input', function() {
        const password = input.value;
        const result = checkPasswordStrength(password);
        
        // 更新進度條
        const percentage = Math.min(result.score, 100);
        let html = `
            <div class="mb-2">
                <div class="progress" style="height: 8px;">
                    <div class="progress-bar bg-${result.color}" 
                         role="progressbar" 
                         style="width: ${percentage}%"
                         aria-valuenow="${percentage}" 
                         aria-valuemin="0" 
                         aria-valuemax="100"></div>
                </div>
            </div>
            <div class="small">
                <div class="text-${result.color} mb-2"><strong>${result.message}</strong></div>
                <div class="row g-2">
                    <div class="col-6">
                        <span class="${result.checks.length ? 'text-success' : 'text-muted'}">
                            <i class="bi bi-${result.checks.length ? 'check-circle-fill' : 'circle'}"></i> 長度 ≥ 8
                        </span>
                    </div>
                    <div class="col-6">
                        <span class="${result.checks.uppercase ? 'text-success' : 'text-muted'}">
                            <i class="bi bi-${result.checks.uppercase ? 'check-circle-fill' : 'circle'}"></i> 大寫字母
                        </span>
                    </div>
                    <div class="col-6">
                        <span class="${result.checks.lowercase ? 'text-success' : 'text-muted'}">
                            <i class="bi bi-${result.checks.lowercase ? 'check-circle-fill' : 'circle'}"></i> 小寫字母
                        </span>
                    </div>
                    <div class="col-6">
                        <span class="${result.checks.digit ? 'text-success' : 'text-muted'}">
                            <i class="bi bi-${result.checks.digit ? 'check-circle-fill' : 'circle'}"></i> 數字
                        </span>
                    </div>
                    <div class="col-6">
                        <span class="${result.checks.special ? 'text-success' : 'text-muted'}">
                            <i class="bi bi-${result.checks.special ? 'check-circle-fill' : 'circle'}"></i> 特殊符號
                        </span>
                    </div>
                </div>
            </div>
        `;
        
        meter.innerHTML = html;
        
        // 存储验证结果供表单提交时使用
        input.dataset.passwordValid = result.valid;
        input.dataset.passwordStrength = result.strength;
    });
}

